staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamic
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/golang/glog"
+	auditregv1alpha1 "k8s.io/api/auditregistration/v1alpha1"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/errors"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit"
+	webhook "k8s.io/apiserver/pkg/util/webhook"
+	bufferedplugin "k8s.io/apiserver/plugin/pkg/audit/buffered"
+	informers "k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+// PluginName is the name reported in error metrics.
+const PluginName = "dynamic"
+
+// Config holds the configuration for the dynamic backend
+type Config struct {
+	// F informer factory for the audit sinks
+	F informers.SharedInformerFactory
+	// KubeClient for event notifications
+	KubeClient clientset.Interface
+	// BufferedConfig is the runtime buffered configuration
+	BufferedConfig *bufferedplugin.BatchConfig
+	// AuthInfoResolverWrapper provides the webhook authentication for in-cluster endpoints
+	AuthInfoResolverWrapper webhook.AuthenticationInfoResolverWrapper
+}
+
+// Backend represents a dynamic backend
+type Backend struct {
+	configuration *auditregv1alpha1.AuditSink
+	delegate      audit.Backend
+	stopChan      chan struct{}
+}
+
+// NewBackend returns a backend that dynamically updates its configuration
+// based on a shared informer.
+func NewBackend(c *Config) (audit.Backend, error) {
+	informer := c.F.Auditregistration().V1alpha1().AuditSinks()
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventSink := &v1core.EventSinkImpl{Interface: c.KubeClient.CoreV1().Events("")}
+	eventBroadcaster.StartRecordingToSink(eventSink)
+
+	scheme := runtime.NewScheme()
+	err := auditregv1alpha1.AddToScheme(scheme)
+	if err != nil {
+		return nil, err
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "apiserver"})
+
+	if c.BufferedConfig == nil {
+		defaultConfig := NewDefaultWebhookBatchConfig()
+		c.BufferedConfig = &defaultConfig
+	}
+
+	cm, err := webhook.NewClientManager(auditinternal.SchemeGroupVersion, auditinternal.AddToScheme)
```

Comment:
what audit event version does the webhook send out?

Reply:
this is an interesting question, should we lock this at v1, or carry through the scheme set by the static webhook?

Reply:
I expected the registration to include the event version the webhook wanted to receive. defaulting to v1 is reasonable, and we can add that as an option if we get a newer version in the future, I suppose